### DISCUSSION TITLE
How to Struct Tag an Inner TOML field?

### DISCUSSION THREAD

**Discussion by hollowaykeanho** [pelletier/go-toml#607] [Questions] [answered]
Is there a way to quickly defines a struct tag pointing an element from the table when parsing the TOML file above directly to `Filesystem` struct? E.g.

```
type Filesystem struct {
    Build string `toml:"<WHAT TO FILL IN HERE FOR Filesystem.Build>"`
}
```
given that the TOML data file that restricts its user to use table instead of using generalized fields such as:
```
Version = "..."  # this is disallowed

[Filesystem]
Build = "..."
Deploy = "..."

[Configs]
...
```

----

I tried a number of possibilities but to no avail (`filesystem.Build`, `Filesystem.Build`, and etc).

Currently, I'm stuck with inline struct definition and feeds it to `Decoder.Decode` function. E.g.:

```
s := struct {
    Filesystem struct { Build string, .... }
} {
    Filesystem: struct { Build string, .... }{},
}

... 
err = decoder.Decode(path, &s)
...
myStruct.Build = s.Filesystem.Build
...
```

**Comment by moorereason** [comment, marked as answer]
I'm assuming you're using v1.  You need a struct to hold the entire document of which the Filesystem is a part.  You don't actually need the struct tags if the field names match.

See this [Go playground example](https://play.golang.org/p/6kiHbV_B1SD) which I'll show here for posterity's sake:

```go
package main

import (
	"fmt"

	toml "github.com/pelletier/go-toml"
)

type Obj struct {
	Version    string
	Filesystem Filesystem
}

type Filesystem struct {
	FooBuild  string `toml:"Build"`
	Deploy string
}

func main() {
	doc := []byte(`Version = "aaa"  # this is disallowed

[Filesystem]
Build = "bbb"
deploy = "ccc"`)

	var v Obj
	err := toml.Unmarshal(doc, &v)
	if err != nil {
		panic(err)
	}

	fmt.Printf("%#v\n", v)
}
```

**Reply by hollowaykeanho** [reply]
Sorry, forgot to mention, I'm was using `v2` since `v2` is near completion.

EDIT:
1. The example is the same as mine where the difference is I'm doing it with inline struct rather than explicitly defining the temporary struct all over the places (based on your example, `Obj`).
2. I'm assuming there is no way to do it directly with the `toml` tag (in your case, without even declaring `Obj`)?

**Reply by moorereason** [reply]
Updated to v2 and anonymous structs:  https://play.golang.org/p/-xa_EiA23Ji

**Reply by hollowaykeanho** [reply]
Let me rephrase my question:

```
package main

import (
	"fmt"

	toml "github.com/pelletier/go-toml/v2"
)

// Filesystem is the app custom operating filesystem.
type Filesystem struct {
	Build    string `toml:"????"`
	Deploy   string `toml:"????"`
}

func main() {
	doc := []byte(`
[Filesystem]
Build = "bbb"
Deploy = "ccc"`)

	v := &Filesystem{}

	err := toml.Unmarshal(doc, &v)
	if err != nil {
		panic(err)
	}

	fmt.Printf("%#v\n", v)
}
```

What to fill inside `????`

It's okay if it is not possible.

**Reply by moorereason** [reply]
Last reply for tonight. Try
```go
v := struct { Filesystem Filesystem }{}
```

**Reply by hollowaykeanho** [reply]
Never mind. I don't think you are getting my question.

I'm convinced that it is not possible unless we use inline struct to wrap around the actual data structure object since you requested me to do `v := struct { Filesystem Filesystem }{}`.

Here's the playground I'm trying to achieve but failed: https://play.golang.org/p/c-wB0lULcs-

Thanks for the help though.

**Reply by moorereason** [reply]
Correct.  It's due to the first thing I said:

> You need a struct to hold the entire document of which the Filesystem is a part.

The data structure you're decoding into has to represent the whole TOML document.  The field tags are irrelevant since the data structures don't match.

Sorry about the back and forth. I was trying to show how to do it correctly when you wanted to know if your way was possible. I guess all my replies were forms of saying "No." 😄
